src/path.c

```diff
@@ -865,6 +933,118 @@ ly_path_compile_predicate_leafref(const struct lysc_node *ctx_node, const struct
     return (ret == LY_ENOTFOUND) ? LY_EVALID : ret;
 }
 
+/**
+ * @brief Compile deref XPath function into ly_path structure.
+ *
+ * @param[in] ctx libyang context.
+ * @param[in] ctx_node Optional context node, mandatory of @p lref.
+ * @param[in] top_ext Extension instance containing the definition of the data being created. It is used to find the top-level
+ * node inside the extension instance instead of a module. Note that this is the case not only if the @p ctx_node is NULL,
+ * but also if the relative path starting in @p ctx_node reaches the document root via double dots.
+ * @param[in] expr Parsed path.
+ * @param[in] oper Oper option (@ref path_oper_options).
+ * @param[in] target Target option (@ref path_target_options).
+ * @param[in] format Format of the path.
+ * @param[in] prefix_data Format-specific data for resolving any prefixes (see ::ly_resolve_prefix).
+ * @param[in,out] tok_idx Index in @p exp, is adjusted.
+ * @param[out] path Compiled path.
+ * @return LY_ERR value.
+ */
+static LY_ERR
+ly_path_compile_deref(const struct ly_ctx *ctx, const struct lysc_node *ctx_node,
+        const struct lysc_ext_instance *top_ext, const struct lyxp_expr *expr, uint16_t oper, uint16_t target,
+        LY_VALUE_FORMAT format, void *prefix_data, uint32_t *tok_idx, struct ly_path **path)
+{
+    LY_ERR ret = LY_SUCCESS;
+    struct lyxp_expr expr2;
+    struct ly_path *arg_path = NULL;
+    struct lysc_node_leaf *deref_leaf_node;
+    struct lysc_type_leafref *lref;
+    char new_path[LYSC_CTX_BUFSIZE] = {'\0'};
+    size_t pos = 0;
+    size_t len = 0;
+    size_t u;
+
+    /* skip deref keyword and '(' '' */
+    *tok_idx += 2;
+
+    /* emebedded functions were already identified count tokens till ')' */
+    while (lyxp_check_token(NULL, expr, *tok_idx, LYXP_TOKEN_PAR2)) {
+        (*tok_idx)++;
+    }
+
+    /* prepare expr representing just deref arg */
+    expr2.tokens = &expr->tokens[2];
+    expr2.tok_pos = &expr->tok_pos[2];
+    expr2.tok_len = &expr->tok_len[2];
+    expr2.repeat = &expr->repeat[2];
+    expr2.used = *tok_idx - 2;
+    expr2.size = expr->size - 2;
+    expr2.expr = expr->expr;
+
+    /* compile just deref arg */
+    ly_path_compile_leafref(ctx, ctx_node, top_ext, &expr2, oper, target, format, prefix_data, &arg_path);
+
+    deref_leaf_node = (struct lysc_node_leaf *)arg_path[LY_ARRAY_COUNT(arg_path) - 1].node;
+    lref = (struct lysc_type_leafref *)deref_leaf_node->type;
+    ly_path_free(ctx, arg_path);
+
+    /* copy leafref path except the last segment */
+    for (u = 0; u < lref->path->used - 2; u++) {
+        len += lref->path->tok_len[u];
+    }
+    strncpy(&new_path[pos], lref->path->expr, len);
+    pos += len;
+
+    /* add '[' */
+    new_path[pos] = '[';
+    ++pos;
+
+    /* add key predicate */
+    u = lref->path->used - 1;
+    strncpy(&new_path[pos], &lref->path->expr[lref->path->tok_pos[u]], lref->path->tok_len[u]);
```

Comment:
I am sorry, I guess I was not clear. What I suggested is to **compile** a leafref with deref by storing the actual dereferenced leafref nodes but I would keep the path expression as it is, with deref.

Reply:
This whole exercise to temporarily create unwinded path is to get the "full path" to the output and not just the last part (as it was before). I didn't find any other way to do it.

To store this dereferenced leafref is also big question. I also thought about this before, but it seems to be useless (there is no real use for it right now). I am not even sure if you mean to store deref node itself or deref node leafref target? I assumed the second one. I also assume to store it as part of first output path element. If I assume wrongly, just let me know how exactly you wish to store this and I will adjust it.

Reply:
Ok, that seems to be strange request. Lets consider this extended example:
```
list l1 {
    key t1;
    leaf t1 {
        type uint8;
    }
    list l2 {
        key t2;
        leaf t2 {
            type uint8;
        }
        list l3 {
            key t3;
            leaf t3 {
                type uint8;
            }
        }
    }
}
leaf r1 {type leafref {path "../l1/t1\";}}
leaf r2 {type leafref {path "deref(../r1)/../l2/t2\";}}
leaf r3 {type leafref {path "../l1[t1 = current()/../r1)/../l2/t2";}}
leaf r4 {type leafref {path "deref(../r2)/../l3/t3";}}
```
Let's compile `r2`. The resulted output path nodes:: `l1`, `l2`, `t2`. `l1` will also have predicate filled in with key set  to `t1`
Let's compile `r3`. The output path nodes: `l1`, `l2`, `t2`. `l1` will also have predicate filled in with key set  to `t1`

If we change it according to your proposal compilation of `r2` and `r3` will not have the same results. Is this really wanted behavior? I think that the output should be equal, shouldn't it?

One additional problem with your proposal is compilation of `r4`. What exact nodes should you expect to have in output path and in which order (I am not 100% sure if order matters or not)? Should it be:
 - `r1`, `l1`, `t1`, `r2`, `l2`, `t2`, `l3`, `t3`
 - `r2`, `l2`, `t2`, `r1`, `l1`, `t1`, `l3`, `t3`
 - `r2`, `r1`, `l1`, `t1`, `l2`, `t2`, `l3`, `t3`

If I compare it to output of `lyxp_atomize()`,  which seems to give similar outputs but in little bit different order:
 - `r4`, `t1`, `r1`, `l1`, `l2`, `t2`, `r2`, `l3`, `t3`

But even using the atoms output, there is still a needs to convert it to `ly_path` structures, which is not very easy way how to do it.

Reply:
> If we change it according to your proposal compilation of r2 and r3 will not have the same results. Is this really wanted behavior? I think that the output should be equal, shouldn't it?

Based on my proposal the output (the compiled nodes) would be different yes. The reason is that when you evaluate each of the paths, different nodes would be traversed, which is what atoms are supposed to represent. And yes, the obvious observation is that you can write equal expressions in many ways with different efficiency. Part of the reason I do not like this `deref()` extension.

> One additional problem with your proposal is compilation of r4. What exact nodes should you expect to have in output path and in which order (I am not 100% sure if order matters or not)? Should it be:

For atoms, the order should not matter.

> But even using the atoms output, there is still a needs to convert it to ly_path structures, which is not very easy way how to do it.

Just another reason why I was not so happy about supporting this but it was all your idea. But you should not rely on `lyxp_atomize()`, use the existing path compilation which should already do everything, it just needs to be adjusted to be able to parse subexpressions (deref parameter) so you can collect all the paths and join them in a single path.

Reply:
Ok, done